rfc/rfc-81/rfc-81.md

```diff
@@ -0,0 +1,108 @@
+
+## Abstract
+Add lightweight LogCompaction to improve the writing performance of the write side, and improve the query performance of the read side (Spark/Presto, etc.) in some scenarios without having to wait for heavy and time-consuming operations such as Compaction or Clustering.
+
+## Background
+The previous LogCompaction mainly merged log files through HoodieMergedLogRecordScanner, and used ExternalSpillableMap internally to achieve record merging, which resulted in performance loss of writing to disk.
+LogCompaction with Merge Sort is introduced to achieve lightweight minor compaction by merging records through N-way streaming of ordered data, thus improving the writing performance of the write side. At the same time, thanks to the ordered data, the query performance on the read side can be improved when the primary key is met.
+
+## Implementation
+### HoodieConfig etc
+Added a new configuration item for the HoodieLogBlock streaming read buffer size. The default value is 10MB.
+#### Flink
+Added a new configuration item to enable LogCompaction. The default value is false.
+Note:
+1. After Flink enables LogCompaction, the default implementation is MergeSort.
+2. Currently, the log format only supports AVRO. After enabling LogCompaction, need to check whether the hoodie.logfile.data.block.format configuration item is correct.
+
+#### Spark
+Added a new configuration item for whether to enable MergeSort in LogCompaction. The default value is false.
+Note:
+1. Currently, the log format only supports AVRO. After LogCompaction turns on MergeSort, you need to check whether the hoodie.logfile.data.block.format configuration item is correct.
+
+### DeltaCommit
+#### Flink
+When LogCompaction is enabled, DeltaCommit sorts the written data by RecordKey to achieve orderly records in LogBlock.
+#### Spark
+When LogCompaction turns on MergeSort, DeltaCommit sorts the written data by RecordKey to achieve orderly records in LogBlock.
+
+### LogCompaction
+The following contents are all about enabling LogCompaction or enabling MergeSort with LogCompaction.
+#### Flink
+Flink has not yet fully implemented the LogCompaction feature, so the operator needs to be implemented:
+1. LogCompactionPlanOperator
+2. LogCompactionOperator
+3. LogCompactionCommitSink
+
+When scanning log files, use the new log scanner: HoodieMergeSortLogRecordScanner to achieve N-way streaming record merging.
+
+#### Spark
+When scanning log files, use the new log scanner: HoodieMergeSortLogRecordScanner to achieve N-way streaming record merging.
+
+### HoodieMergeSortLogRecordScanner
+Implement a min-heap. The heap node is a HoodieLogBlock object. The heap node comparison uses HoodieRecord#RecrodKey.
+When traversing the scanner record iterator, return to the top node of the heap and call the HoodieLogBlock object to return the record.
+
+### HoodieDataBlock
+Added a new streaming read abstract method to help HoodieMergeSortLogRecordScanner avoid OOM when reading N LogBlocks.
```

Comment:
are we worried about parquet memory usage? for avro records, there should not be much memory overhead, right.

Reply:
> are we worried about parquet memory usage? for avro records, there should not be much memory overhead, right.

We do not need to worry about the memory overhead of Parquet, because LogCompaction does not involve Parquet.

The reason why we care about the memory consumption of avro records here is because N-way streaming record merging requires opening all log data blocks